core/src/main/java/com/linecorp/armeria/client/limit/DefaultConcurrencyLimit.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright 2021 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.limit;
+
+import static com.linecorp.armeria.client.limit.ConcurrencyLimitBuilder.noLimitFuture;
+
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Predicate;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+import com.spotify.futures.CompletableFutures;
+
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.client.UnprocessedRequestException;
+import com.linecorp.armeria.common.util.SafeCloseable;
+
+/**
+ * Concurrency settings that limits the concurrent number of active requests.
+ */
+final class DefaultConcurrencyLimit implements ConcurrencyLimit {
+
+    private final Predicate<? super ClientRequestContext> predicate;
+    private final int maxConcurrency;
+    private final int maxPendingAcquisitions;
+    private final long timeoutMillis;
+
+    private final Queue<PendingAcquisition> pendingAcquisitions = new ConcurrentLinkedQueue<>();
+    private final AtomicInteger acquiredPermits = new AtomicInteger();
+
+    DefaultConcurrencyLimit(Predicate<? super ClientRequestContext> predicate,
+                            int maxConcurrency, int maxPendingAcquisitions, long timeoutMillis) {
+        this.predicate = predicate;
+        this.maxConcurrency = maxConcurrency;
+        this.maxPendingAcquisitions = maxPendingAcquisitions;
+        this.timeoutMillis = timeoutMillis;
+    }
+
+    @VisibleForTesting
+    int acquiredPermits() {
+        return acquiredPermits.get();
+    }
+
+    @VisibleForTesting
+    int availablePermits() {
+        return maxConcurrency - acquiredPermits.get();
+    }
+
+    @Override
+    public CompletableFuture<SafeCloseable> acquire(ClientRequestContext ctx) {
+        if (!predicate.test(ctx)) {
+            return noLimitFuture;
+        }
+
```

Comment:
Is this a valid optimization to avoid the queue for stable concurrency?

```java
// Try acquiring a permit
if (acquired.incrementAndGet() <= maxConcurrency && pendingAcquisitions.isEmpty()) {
  // Got a valid permit
  return completedFuture(new Permit())
}
// Could not get a valid permit. First return the invalid one
acquired.decrementAndGet();

if (numPendingAcquisitions.incrementAndGet() <= maxPendingAcquisitions) {
  // Too many pending acquisitions already don't try waiting
  numPendingAcquisitions.decrementAndGet()
  return exceptionalFuture(Exceeded)
}

drain();

Reply:
That's a good point.
To strictly guarantee for requests to get through the client decorator in order, we need to roll back to use the queue. 
But, even without the `ConcurrentLimitingClient`, requests are not sent in order because the order is changed by DNS resolving and other stuff in the `HttpClientDelegate`. So I thought this was okay.
Let me leave a comment about it.
And please tell me if you think we need to guarantee the order, then I will roll back to the previous implementation. 😄

Reply:
Yes it is hard to provide the exact order of requests. The current implementation is reasonable.

Reply:
On second thought, we could use the `ConcurrencyLimit` other than `ConcurrentLimitingClient`. So probably, it's better to guarantee the FIFO? 🤔

Reply:
I also thought about that. If the code is under common package, we should think about that. But the current API looks to belong to `client.limit`

Reply:
Yeah my rule of thumb is usually because input order can't be guaranteed, output order can't be guaranteed. If limits and fairness within a reasonable window is achieved, it's good for almost all cases.